lib/validators/json-schema-next.js

```diff
@@ -0,0 +1,173 @@
+const Ajv = require('ajv');
+const tv4 = require('tv4');
+
+const metaSchemaV6 = require('ajv/lib/refs/json-schema-draft-06.json');
+const metaSchemaV7 = require('ajv/lib/refs/json-schema-draft-07.json');
+
+const metaSchemaV4 = require('../meta-schema-v4');
+const metaSchemaV3 = require('../meta-schema-v3');
+const errors = require('../errors');
+const parseJson = require('../utils/parseJson');
+
+const SCHEMA_VERSIONS = {
+  draftV3: 'http://json-schema.org/draft-03/schema',
+  draftV4: 'http://json-schema.org/draft-04/schema',
+  draftV6: 'http://json-schema.org/draft-06/schema',
+  draftV7: 'http://json-schema.org/draft-07/schema'
+};
+
+const META_SCHEMA = {
+  draftV3: metaSchemaV3,
+  draftV4: metaSchemaV4,
+  draftV6: metaSchemaV6,
+  draftV7: metaSchemaV7
+};
+
+/**
+ * Returns a JSON Schema Draft version of the given JSON Schema.
+ */
+const getSchemaVersion = (jsonSchema) => {
+  const currentVersion = jsonSchema.$schema && jsonSchema.$schema;
+  return Object.keys(SCHEMA_VERSIONS).find((version) => {
+    const jsonSchemaAnnotation = SCHEMA_VERSIONS[version];
+    return currentVersion && currentVersion.includes(jsonSchemaAnnotation);
+  });
+};
+
+/**
+ * @deprecate
+ * Attempts to resolve a schema version for a JSON Schema
+ * without the explicit version.
+ */
+const getImplicitSchemaVersion = (jsonSchema) => {
+  const [schemaVersion] = [
+    ['draftV3', metaSchemaV3],
+    ['draftV4', metaSchemaV4]
+  ].find(([_, metaSchema]) => {
+    tv4.reset();
+    tv4.addSchema('', metaSchema);
+    tv4.addSchema(metaSchema.$schema, metaSchema);
+    const validationResult = tv4.validateResult(jsonSchema, metaSchema);
+    return validationResult.valid;
+  }) || [null, null];
+
+  return schemaVersion;
+};
+
+class JsonSchemaValidator {
+  constructor(jsonSchema) {
+    this.jsonSchema = jsonSchema;
+    this.jsonSchemaVersion =
```

Comment:
In Draft 6 and onwards, `true` and `false` are valid schemas (https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.4.3.2). I think it would be worth testing these two values in regards to the version detection. I'd gather it would end up being draft 3/4 as implicit schema and then become an invalid schema.

Reply:
Perhaps we would want to default it to Draft 7 under those circumstances since the behaviour would be same in draft 6, 7 and 2019-09

Reply:
Inferring of JSON Schema Draft 7 implemented in https://github.com/apiaryio/gavel.js/pull/355/commits/593ea41dc86af6b935f7e45645a8bf13f728944f.